### ISSUE TITLE
Anchors – setting object on a null property on attached component doesn't trigger update 

### ISSUE THREAD

**Issue by marcalexiei** [ractivejs/ractive#3278]
## Description:
Follow up #3276

1. Set a property to `null` or `undefined` in `data`.
2. Create a computed property which performs a get on a nested property of above property.
3. Print the result in the template.
Computed property doesn't update after "deferred" `set`. 

## Versions affected:
* 1.0.1-build-3

## Platforms affected:
All

## Reproduction:
1. Load this [fiddle](https://jsfiddle.net/nhq5mb4L/)
2. After ~500ms the text "foo.bar is not defined" should become "foo.bar is defined" but nothing happens.

```javascript
Ractive.defaults.isolated = false;

const r = window.r = new Ractive({
  template: `
  <h1>A Anchrome try</h1>
 	<p>@shared.sharedData inside root component: {{@shared.sharedData}}</p>
	<# mount-point />
  `,
  data: {}
});

r.render('#main').then(async () => {
  const MyAttachedComponent = Ractive.extend({
    template: `
     <h2>B Attached component</h2>

     <p>
       {{#if isFooBar}}
          foo.bar is defined
       {{else}}
          foo.bar is not defined
       {{/if}}
     </p>
    `,

    data() {
      return {
        foo: undefined,
      };
    },

    computed: {
      isFooBar() {
        const check = this.get('foo.bar');
        return check;
      }
    },

    onrender() {
      setTimeout(() => {
        this.set({
          foo: { bar: 'pippo' },
        });
      }, 500);
    },
  });

  await r.attachChild(new MyAttachedComponent(), { target: 'mount-point' });
});
```

**Comment by dagnelies** [comment]
OMG ...I never used so many fancy stuff. Honestly, I wonder if there is no way to simplify your stuff ;) :P 

The fiddle is also not the right one, you copied it from the previous ticket.

The fixed one is here: https://jsfiddle.net/w4ruo0n5/1/

Basically, use `Ractive.set('keypath', value)` or `Ractive.set({'keypath1': value1, 'keypath2': value2, ...})`.

However, `Ractive.set(myObject)` is simply wrong according to the API.

**Comment by marcalexiei** [comment]
> OMG ...I never used so many fancy stuff. Honestly, I wonder if there is no way to simplify your stuff ;) :P

This fiddle is a simplified version of a more complex scenario in a private application that I'm trying to rewrite using `attachChild`.

> The fiddle is also not the right one, you copied it from the previous ticket.

Yes, I have updated the link in the initial comment. Thanks.

> The fixed one is here: https://jsfiddle.net/w4ruo0n5/1/

1. I don't see why use arrow function using `var self = this` before... since `this` is keep inside arrow function.
2. template is still not working, without `attachChild` everything works as expected.

> However, Ractive.set(myObject) is simply wrong according to the API.

> Basically, use Ractive.set('keypath', value) or Ractive.set({'keypath1': value1, 'keypath2': value2, ...}).

`Ractive.set({'keypath1': value1, 'keypath2': value2, ...})` isn't the same of `Ractive.set(myObject)`?

From https://ractive.js.org/api/#ractiveset:
`ractive.set(map[, options])`

**Comment by dagnelies** [comment]
Sorry, I saved the wrong one in the middle of toying around, the "self=" were useless at this point.

...and my previous comment was full of nonsense, I must be tired. Of course the `set` was OK was correctly used.

That said, I don't know exactly why the provided example does not work. Here are two examples that do work:

`set('foo.bar', 'pippo')` => https://jsfiddle.net/w4ruo0n5/2/
`isolated = false` => https://jsfiddle.net/w4ruo0n5/3/

I guess the root of the problem has not really to do with you, but rather the contrived way of how `Ractive.defaults.isolated = false;` deals with the keypath resolution. Even as a long term user of ractive, I have given up trying to know "where" non-isolated data lands. If you do complex stuff, I strongly advise you to avoid this flag, which is rarely necessary. In my experience, it only leads to subbtle issues down the road.

**Comment by dagnelies** [comment]
Perhaps @evs-chris can shed more light on this ...I also don't know if his previous fix is already in the latest build, it might be related.

**Comment by dagnelies** [comment]
Here is one more example that works (by initializing the data in the child component)

https://jsfiddle.net/w4ruo0n5/5/

...the more I toy around, the more I conclude that `isolated = false` is buggy ...it was always a source of bugs.

Especially in the original case, I wonder where it lands though (neither in the parent, nor in the child):

https://jsfiddle.net/w4ruo0n5/6/

**Comment by marcalexiei** [comment]
The problem is that we start the app with Ractive 0.7 and we are updating it constantly but some section needs a total recode if we use `isolated = true`.

**Comment by evs-chris** [comment]
I've got this fixed on edge (when travis finishes travising), but if you're doing what I think you're doing, it will probably still not work as you need it to. I _think_ what you're after is an attached instance that automatically picks up _certain_ data from its parent environment, but with this change, I don't think it will any more. I _think_ the model attachment will need to be updated to handle explicitly missing data e.g. `{ foo: undefined }` as opposed to implicitly missing data `{}`, but I'll wait on confirmation.

**Comment by evs-chris** [comment]
Just as a comment on isolation being buggy, it's not so much that it's buggy as it's implicitly ambiguous, and I'm sure everyone here knows that ambiguity is the father of ractive pain and the cousin of heartburn :smile:.

**Comment by marcalexiei** [comment]
HI, I tried now on my application using `1.1.0-build-2` and it seems to be working.
 
Regarding
> I think what you're after is an attached instance that automatically picks up certain data from its parent environment
 
In our application we are try to use `attachChild` to render the "PageBody" component. This component performs an API call and set the relevant data on itself (simulate in the fiddle using `setTimeout`). 
From the example `<MyAttachedComponent/>` is getting / setting data only on itself.
If we need data from parent component, which is the root component,  we are using `@.root.get` or `@shared.property` and hopefully one day we will set `Ractive.defaults.isolated` to `true`!
 
Anyway thanks for the fix!
 
P.S.: (nearly off topic) `warnAboutAmbiguity` property is awesome to track  to down when data is not resolved in current context or component.
